### ISSUE TITLE
Unit-тесты для сервиса по определению типа волос с помощью разработанной модели

### ISSUE THREAD

**Issue by Linasthsmb** [Fairies-Pixels/curlylab-android#66]
# Тестирование API определения пористости волос

Для проверки корректности работы сервиса были реализованы unit-тесты с использованием pytest и FastAPI TestClient.
Тесты покрывают основную бизнес-логику сервиса, обработку изображений и работу HTTP-endpoint.

### predict_porosity

Проверяется корректность работы функции предсказания пористости волос.

Тестируются следующие сценарии:

* Функция возвращает строку класса пористости
  Возможные значения:

  * low
  * medium
  * high

* Модель действительно вызывается внутри функции predict_porosity.

* Корректное сопоставление индекса класса модели и текстового результата:

  * 0 → low
  * 1 → high
  * 2 → medium

---

### Преобразование изображения

Проверяется корректность работы porosity_transform.

Тесты проверяют:

* преобразование изображения в torch.Tensor
* наличие трёх каналов (RGB)

---

### Endpoint /analyze

Проверяется корректная обработка изображения API.

Тесты проверяют:

* успешную отправку изображения
* корректный HTTP-статус ответа (200)
* корректный JSON-ответ

Пример ответа API:

{
  "porosity": "low"
}

---

### Проверка типа файла

Endpoint должен отклонять файлы, которые не являются изображениями.

Если отправить файл другого формата (например txt), сервис должен:

* вернуть статус 400
* вернуть сообщение об ошибке

Файл должен быть изображением JPG или PNG

---

## Проверка взаимодействия компонентов

Отдельно проверяется, что endpoint /analyze действительно вызывает функцию predict_porosity.

Для этого используется mock-объект, который позволяет:

* изолировать тесты от реальной модели
* проверить факт вызова функции предсказания.

---

## Вспомогательные элементы тестов

Для тестирования используется функция create_test_image, которая:

* генерирует тестовое изображение через Python Imaging Library
* сохраняет его в BytesIO
* отправляет изображение в API как файл.

---

## Покрываемые компоненты

Тесты проверяют:

* функцию predict_porosity
* преобразование изображений porosity_transform
* endpoint /analyze
* обработку ошибок
* взаимодействие API и модели

# Тестирование загрузки модели

Для проверки корректности загрузки модели были реализованы unit-тесты для модуля models_loader.
Тесты проверяют корректность инициализации модели, загрузку весов и обработку ошибок.

Тестирование выполняется с использованием pytest и mock-объектов.

---

# Позитивные тесты

### Создание модели

Проверяется корректность создания архитектуры модели.

Тесты проверяют:

* вызов функции create_model
* корректные параметры создания модели

Параметры архитектуры:

* модель: convnext_base
* pretrained=False
* num_classes=3

Модель создается с помощью библиотеки
timm.

---

### Загрузка весов модели

Проверяется корректность загрузки весов из checkpoint.

Тест проверяет:

* вызов torch.load
* передачу model_state_dict в load_state_dict
* успешную инициализацию модели

Используется библиотека
PyTorch.

---

### Перевод модели в режим inference

Проверяется, что после загрузки весов модель переводится в режим:

model.eval()

Это необходимо для:

* отключения dropout
* корректной работы batch normalization
* использования модели в режиме предсказания.

---

# Негативные тесты

### Отсутствие файла модели

Если файл модели отсутствует, функция должна выбрасывать исключение:

FileNotFoundError

Тест проверяет:

* наличие проверки os.path.exists
* корректную обработку отсутствующего файла модели.

---

# Проверка взаимодействия компонентов

Для изоляции тестов используются mock-объекты, позволяющие:

* не загружать реальную модель
* не использовать реальные веса
* ускорить выполнение тестов

Мокаются следующие компоненты:

* os.path.exists
* timm.create_model
* torch.load
* model.load_state_dict

---

# Покрываемые компоненты

Тесты проверяют:

* функцию load_porosity_model
* создание архитектуры модели
* загрузку весов
* перевод модели в режим eval
* обработку отсутствующего файла модели

---

В совокупности тесты покрывают:

* загрузку модели
* обработку изображений
* работу API

что обеспечивает стабильность работы сервиса и предотвращает регрессии при изменении кода.
